thewalrus/quantum.py

```diff
@@ -1,1311 +1,1415 @@
+
+    Note that if the covariance matrices correspond to pure states this
+    function reduces to the modulus square of the overlap of their state vectors.
+    For the derivation see  `'Quantum Fidelity for Arbitrary Gaussian States', Banchi et al. <10.1103/PhysRevLett.115.260501>`_.
+
+    Args:
+        mu1 (array): vector of means of the first state
+        cov1 (array): covariance matrix of the first state
+        mu2 (array): vector of means of the second state
+        cov2 (array): covariance matrix of the second state
+        hbar (float): value of hbar in the uncertainty relation
+        rtol (float): the relative tolerance parameter used in `np.allclose`
+        atol (float): the absolute tolerance parameter used in `np.allclose`
+
+    Returns:
+        (float): value of the fidelity between the two states
+    """
+
+    n0, n1 = cov1.shape
+    m0, m1 = cov2.shape
+    (l0,) = mu1.shape
+    (l1,) = mu1.shape
+    if not n0 == n1 == m0 == m1 == l0 == l1:
+        raise ValueError("The inputs have incompatible shapes")
+
+    v1 = cov1 / hbar
+    v2 = cov2 / hbar
+    deltar = (mu1 - mu2) / np.sqrt(hbar / 2)
+    n0, n1 = cov1.shape
+    n = n0 // 2
+    W = sympmat(n)
+
+    si12 = np.linalg.inv(v1 + v2)
+    vaux = W.T @ si12 @ (0.25 * W + v2 @ W @ v1)
+    p1 = vaux @ W
+    p1 = p1 @ p1
+    p1 = np.identity(2 * n) + 0.25 * np.linalg.inv(p1)
+    if np.allclose(p1, 0, rtol=rtol, atol=atol):
+        p1 = np.zeros_like(p1)
+    else:
+        p1 = sqrtm(p1)
+    p1 = 2 * (p1 + np.identity(2 * n))
+    p1 = p1 @ vaux
+    f = np.sqrt(np.linalg.det(si12) * np.linalg.det(p1)) * np.exp(
+        -0.25 * deltar @ si12 @ deltar
+    )
+    return f
+
+def gen_normal_ordered_complex_cov(cov, hbar=2):
+    r"""Calculates the normal ordered covariance matrix in the complex basis.
+
+    Args:
+        cov (array): xp-covariance matrix.
+        hbar (float): value of hbar in the uncertainty relation.
+
+    Returns:
+        (array): covariance matrix in the creation/annihilation operator basis.
+    """
+
+    n, _ = cov.shape
+    n_modes = n // 2
+    cov = cov / (hbar / 2)
+    A = cov[:n_modes, :n_modes]
+    B = cov[:n_modes, n_modes:]
+    C = cov[n_modes:, n_modes:]
+    N = 0.25 * (A + C + 1j * (B - B.T) - 2 * np.identity(n_modes))
+    M = 0.25 * (A - C + 1j * (B + B.T))
+    mat = np.block([[M.conj(), N], [N.T, M]])
+    return mat
+
+
+def normal_ordered_expectation(mu, cov, rpt, hbar=2):
+    r"""Calculates the expectation value of the normal ordered product
+    :math:`\prod_{i=0}^{N-1} a_i^{\dagger n_i} \prod_{j=0}^{N-1} a_j^{m_j}` with respect to an N-mode Gaussian state,
+    where :math:`\text{rpt}=(n_0, n_1, \ldots, n_{N-1}, m_0, m_1, \ldots, m_{N-1})`.
+
+    Args:
+        mu (array): length-:math:`2N` means vector in xp-ordering.
+        cov (array): :math:`2N\times 2N` covariance matrix in xp-ordering.
+        rpt (list): integers specifying the terms to calculate.

```

Comment:
What does `rpt` do here? It's not too clear to me

Reply:
The normal ordered product is defined as :math:`\prod_{i=0}^{N-1} a_i^{\dagger n_i} \prod_{j=0}^{N-1} a_j^{m_j}` where the user has to provide the integers n_i and m_j. These are passed in `rpt` via  :math:`\text{rpt}=(n_0, n_1, \ldots, n_{N-1}, m_0, m_1, \ldots, m_{N-1})`

Reply:
Ohh! Worth adding this to the docstring I think

Reply:
It is already in the docstring :)

Reply:
🤦 it was too early when I reviewed, and too late just now when I commented!